File: solutions/java/code-carl/ch07_backtracking/T12_LC0047_PermuteUnique.java
/**
 * Practice template for:
 * - LeetCode 47. Permutations II
 *   https://leetcode.cn/problems/permutations-ii/
 */
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class T12_LC0047_PermuteUnique {
    List<List<Integer>> res = new ArrayList<>();
    List<Integer> path = new ArrayList<>();
    boolean[] used;

    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        used = new boolean[nums.length];
        backtracking(nums);
        return res;
    }

    public void backtracking(int[] nums) {
        if (path.size() == nums.length) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (used[i] ||
                    i > 0 && nums[i] == nums[i - 1] && !used[i - 1]) {
                continue;
            }
            path.add(nums[i]);
            used[i] = true;
            backtracking(nums);
            path.remove(path.size() - 1);
            used[i] = false;
        }
    }

}

File: solutions/code-carl/ch07_backtracking/t12_lc0047_permute_unique.py
"""LeetCode 47. Permutations II
https://leetcode.cn/problems/permutations-ii/
"""

from typing import List


class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        path: List[int] = []
        ans: List[List[int]] = []
        used: List[bool] = [False] * len(nums)

        def backtracking(nums: List[int]) -> None:
            if len(path) == len(nums):
                ans.append(path[:])
                return

            for i in range(0, len(nums)):
                if used[i] or i > 0 and nums[i] == nums[i - 1] and not used[i - 1]:
                    continue

                path.append(nums[i])
                used[i] = True
                backtracking(nums)
                path.pop()
                used[i] = False

        nums.sort()
        backtracking(nums)
        return ans


if __name__ == "__main__":
    solution = Solution()
    result = {tuple(x) for x in solution.permuteUnique([1, 1, 2])}

    expected = {(1, 1, 2), (1, 2, 1), (2, 1, 1)}
    assert result == expected

    solution = Solution()
    result = {tuple(x) for x in solution.permuteUnique([1, 2, 3])}
    expected = {
        (1, 2, 3),
        (1, 3, 2),
        (2, 1, 3),
        (2, 3, 1),
        (3, 1, 2),
        (3, 2, 1),
    }
    assert result == expected

    print("All tests passed.")
